processor_labels: Add a documentation for processor_labels (#1371)
fluent/fluent-bit-docs: pipeline/processors/labels.md

# Labels

The **labels** processor allows you to manipulate the labels of Metrics.

Similar to the functionality exposed by filters, this processor presents a enriching/modifying mechanism to perform such operations for labels manipulation.
The most significant difference is that processors perform better than filters, and when chaining them, there are no encoding/decoding performance penalties.

> Note that processors and this specific component can only be enabled using the new YAML configuration format. Classic mode configuration format doesn't support processors.

Reviewer: ```suggestion
Note that processors and this specific component can only be enabled using the new YAML configuration format.
Classic mode configuration format doesn't support processors.
```

## Configuration Parameters

| Key         | Description |
| :---------- | :--- |

> | update  | Update an existing key with a value into metrics. The key-value pair is required. |

Reviewer: What happens if it does not exist? Failure?

Author: No, just dismissed the operation.

Reviewer: Can you clarify that in the docs? Some things seem to error and some ignore so useful to know.

> | insert  | Insert a new key with a value into metrics. The key-value pair is required. |

Reviewer: What happens if it exists?

Author: Just ignored in that case on https://github.com/fluent/fluent-bit/pull/8824 patch.

> | upsert  | Upsert a specific key with a value, the `upsert` operation will try to update the value of the key. If the key does not exist, the key will be created. The key-value pair is required. |

Reviewer: This seems like the option to always use :)

Author: Yeah, but it's both of effective whether existing the key or not. :)

> | delete  | Delete a key from the labels of metrics. The key-value pair is required. |

Reviewer: What happens if it does not exist?

Author: Just ignored and dismissed in that case.

> | hash    | Replace the key value with a hash generated by the SHA-256 algorithm from the specified label name, the binary value generated is finally set as an hex string representation. |

Reviewer: Weirdly specific option - can we change the hash algorithm and other stuff in the future?

Author: Not sure. But I suppose that we shouldn't do this. Because there is no reason to use weaker hash algorithms such as SHA-1 or MD5.

Author: This should be used for anonymization of labels' values.

# Labels

> The **labels** processor allows you to manipulate the labels of Metrics.
> 
> Similar to the functionality exposed by filters, this processor presents a enriching/modifying mechanism to perform such operations for labels manipulation.
> The most significant difference is that processors perform better than filters, and when chaining them, there are no encoding/decoding performance penalties.
> 
> Note that processors and this specific component can only be enabled using the new YAML configuration format.
> Classic mode configuration format doesn't support processors.

Reviewer 2: Adding some formatting and grammar suggestions for this, along with some line wraps to mesh with style guidance we now have in place.

```suggestion
The **labels** processor lets you manipulate the labels of metrics.

Similar to filters, this processor presents a enriching/modifying mechanism to
perform operations for labels manipulation. The most significant difference is
that processors perform better than filters, and when chaining them there are no
encoding or decoding performance penalties.

{% hint style="info" %}
**Note:** Processors and this specific component can be
enabled only by using the YAML configuration format. Classic mode configuration
format doesn't support processors.
{% endhint %}
```

## Configuration Parameters

> | Key         | Description |
> | :---------- | :--- |
> | update  | Update an existing key with a value into metrics. The key-value pair is required. If the specified key does not exist, the operation is silently failed and no effect. |
> | insert  | Insert a new key with a value into metrics. The key-value pair is required. |

Reviewer 2: ```suggestion
| Key    | Description |
| :----- | :---------- |
| update | Update an existing key with a value into metrics. The key/value pair is required. If the specified key doesn't exist, the operation silently fails and has no effect. |
| insert | Insert a new key with a value into metrics. The key/value pair is required. |
```

| upsert  | Upsert a specific key with a value, the `upsert` operation will try to update the value of the key. If the key does not exist, the key will be created. The key-value pair is required. |

> | delete  | Delete a key from the labels of metrics. The key-value pair is required. If the specified key does not exist, the operation is silently failed and no effect. |
> | hash    | Replace the key value with a hash generated by the SHA-256 algorithm from the specified label name, the binary value generated is finally set as an hex string representation. |

Reviewer 2: ```suggestion
| delete | Delete a key from the labels of metrics. The key/value pair is required. If the specified key doesn't exist, the operation silently fails and has no effect. |
| hash   | Replace the key value with a hash generated by the SHA-256 algorithm from the specified label name. The generated binary value is set as a hex string. |
```

> 

Reviewer 2: Removing this unneeded blank line.

```suggestion
```

#### Update example

Change the value of the `name` to `fluentbit`:

```yaml
pipeline:
  inputs:
    - name: fluentbit_metrics
      processors:
        metrics:
          - name: labels
            update: name fluentbit
  outputs:
    - name : stdout
      match: '*'
```

#### Insert example

> The following example appends the key `agent` with the value `fluentbit` as the label of metrics.

Reviewer 2: ```suggestion
The following example appends the key `agent` with the value `fluentbit` as the label
of metrics:
```
